VultisigApp/VultisigApp/Services/CryptoPriceService.swift

```diff
@@ -40,39 +66,88 @@ public class CryptoPriceService: ObservableObject {
                 network: chain.coingeckoId,
                 addresses: [contractAddress])
             )
+            
             if let response = response.data.first {
                 let priceRate = response.attributes.price_usd.flatMap { Double($0) }
+                
+                // Cache the fetched price
+                if let priceRate = priceRate {
+                    cacheTokens[cacheKey] = (data: [contractAddress: priceRate], timestamp: Date())
+                }
+                
                 return (response.attributes.image_url, response.attributes.coingecko_coin_id, priceRate)
             }
-
         } catch {
             print(error.localizedDescription)
             return (image_url: nil, coingecko_coin_id: nil, price_usd: nil)
         }
+        
         return (image_url: nil, coingecko_coin_id: nil, price_usd: nil)
     }
     
-    func fetchCoingeckoId(chain: Chain, addresses: [String]) async throws -> [String?] {
-        struct Response: Codable {
-            struct Data: Codable {
-                struct Attributes: Codable {
-                    let coingecko_coin_id: String?
+    func getTokenPrice(coin: Coin) async -> Double {
+        let cacheKey = getCacheTokenKey(contractAddress: coin.contractAddress, chain: coin.chain)
+        
+        if let cacheEntry = await getCachedTokenPrice(contractAddress: coin.contractAddress, chain: coin.chain) {
+            print("getTokenPrice > from cache \(cacheKey)")
+            return cacheEntry
+        } else {
+            let prices = await fetchCoingeckoTokenPrice(contractAddresses: [coin.contractAddress], chain: coin.chain)
+            return prices[cacheKey] ?? .zero
+        }
+    }
+    
+    private func fetchCoingeckoTokenPrice(contractAddresses: [String], chain: Chain) async -> [String: Double] {
+        var tokenPrices: [String: Double] = [:]
+        let fiat = SettingsCurrency.current.rawValue.lowercased()
+        do {
+            // Create a cache key for each contract address individually
+            for address in contractAddresses {
+                let cacheKey = getCacheTokenKey(contractAddress: address, chain: chain)
+                //if let cacheEntry = cacheTokens[cacheKey], isCacheValid(for: cacheKey) {
+                if let cacheEntry = await getCachedTokenPrice(contractAddress: address, chain: chain) {
+                    print("Price from cache tokens")
+                    print("fetchCoingeckoTokenPrice > cacheKey: \(cacheKey)")
+                    tokenPrices[address] = cacheEntry
+                    continue
                 }
-                let attributes: Attributes
             }
-            let data: [Data]
+            
+            // If no cache entry is found, fetch the prices for all contract addresses
+            let urlString = Endpoint.fetchTokenPrice(network: chain.name, addresses: contractAddresses, fiat: fiat)
```

Comment:
It looks like even the price is in cache , this will still fire a request to get the token prices